Tools/peg_generator/peg_parser/pegen.c

```diff
@@ -595,8 +601,15 @@ run_parser_from_file(const char *filename, void *(start_rule_func)(Parser *), in
     tok->filename = filename_ob;
     filename_ob = NULL;
 
-    result = run_parser(tok, start_rule_func, mode, FILE_INPUT, keywords, n_keyword_lists);
+    Parser *p = Parser_New(tok, parse_func, FILE_INPUT, arena);
+    if (p == NULL) {
+        PyTokenizer_Free(tok);
+        goto error;
+    }
+
+    result = run_parser(p, mode);
 
+    Parser_Free(p);
     PyTokenizer_Free(tok);
 
 error:
```

Comment:
Perhaps a simpler way to keep track of the necessary free/close calls is to initialize `tok` and `p` to `NULL` and then move the error label before the `Parser_Free` call, and do the free/close calls only if the point isn't `NULL`. This makes things more robust when you change the logic.

Reply:
I've got a whole lot of work locally, which fixes this in another way. I'm gonna fix the nits and push a new PR, like you proposed. We can discuss this over there.